lib/fog/compute/ecloud.rb

```diff
@@ -0,0 +1,822 @@
+            @password              = options[:ecloud_password]
+            if @username.nil? || @password.nil?
+              raise ArgumentError, "No credentials (cloud auth, or basic auth) passed!"
+            end
+          else
+            @hmac = Fog::HMAC.new("sha256", @private_key)
+          end
+        end
+
+        def request(params)
+          # Convert the uri to a URI if it's a string.
+          if params[:uri].is_a?(String)
+            params[:uri] = URI.parse(@host + params[:uri])
+          end
+          host_url = "#{params[:uri].scheme}://#{params[:uri].host}#{params[:uri].port ? ":#{params[:uri].port}" : ""}"
+
+          # Hash connections on the host_url ... There"s nothing to say we won"t get URI"s that go to
+          # different hosts.
+          @connections[host_url] ||= Fog::XML::Connection.new(host_url, @persistent, @connection_options)
+
+          # Set headers to an empty hash if none are set.
+          headers = set_extra_headers_for(params) || set_extra_headers_for({})
+
+          # Make the request
+          options = {
+            :expects => (params[:expects] || 200),
+            :method  => params[:method] || "GET",
+            :path    => params[:uri].path + "#{"?#{params[:uri].query}" if params[:uri].query}",
+            :headers => headers
+          }
+          unless params[:body].nil? || params[:body].empty?
+            options.merge!(:body => params[:body])
+          end
+          response = @connections[host_url].request(options)
+          # Parse the response body into a hash
+          unless response.body.empty?
+            if params[:parse]
+              document = Fog::ToHashDocument.new
+              parser = Nokogiri::XML::SAX::PushParser.new(document)
+              parser << response.body
+              parser.finish
+
+              response.body = document.body
+            end
+          end
+
+          response
+        end
+
+        private
+
+        # if Authorization and x-tmrk-authorization are used, the x-tmrk-authorization takes precendence.
+        def set_extra_headers_for(params)
+          params[:headers] = {
+            "x-tmrk-version" => @version,
+            "Date"           => Time.now.utc.strftime("%a, %d %b %Y %H:%M:%S GMT"),
+          }.merge(params[:headers] || {})
+          if params[:method]=="POST" || params[:method]=="PUT"
+            params[:headers].merge!("Content-Type" => "application/xml") unless params[:headers]["Content-Type"]
+            params[:headers].merge!("Accept" => "application/xml")
+          end
+          unless params[:body].nil? || params[:body].empty?
+            params[:headers].merge!({ "x-tmrk-contenthash" => "Sha256 #{Base64.encode64(Digest::SHA2.digest(params[:body].to_s)).chomp}" })
+          end
+          if @authentication_method == :basic_auth
+            params[:headers].merge!({ "Authorization" => "Basic #{Base64.encode64(@username+":"+@password).delete("\r\n")}" })
```

Comment:
Redundant curly braces around a hash parameter.<br>Surrounding space missing for operator '+'.<br>Prefer single-quoted strings inside interpolations.